### PR TITLE
Fix empty path segments in Data Path transformations

### LINKED ISSUE #449: [BUG] Delta RS read fails if you have an empty path segment
**Describe the bug**
```
path = S3Path.from_hdfs_path("s3a://bucket//path")
```
using that to with `load` or `history`, causes: `OSError: Generic URL error: Path: Path "//tables/return_optimizer/sku" contained empty path segment`

### PR DESCRIPTION
**Pull request by Andre-Lx-Costa**
Fixes #449 

## Scope

Implemented:
 - Fixed path transformations between S3 DataPath and paths to handle empty path segments
 
## Checklist

- [x] GitHub issue exists for this change.
- [x] Unit tests added and they pass.
- [X] Pylint 10.0/10.0 without bloating `.pylintrc` with exceptions.
- [X] Review requested on `latest` commit.

### CODE REVIEW

File: adapta/storage/models/aws.py [hunk 1]
```diff
@@ -36,7 +36,7 @@ def to_uri(self) -> str:
         if not self.bucket or not self.path:
             raise ValueError("Bucket and path must be defined")
 
-        return f"s3://{self.bucket}/{self.path}"
+        return f"s3a://{self.bucket.rstrip('/')}/{self.path}"
```
[the remark this answers is not in the document]
Comment (george-zubrienko) [reply adapta/storage/models/aws.py:39]: same for path

[the remark this answers is not in the document]
Comment (Andre-Lx-Costa) [reply adapta/storage/models/aws.py]: Added `__post_init__` with a regex expression to check all valid cases for s3 data paths as suggested in [cbf15e9](https://github.com/SneaksAndData/adapta/pull/453/commits/cbf15e9e7b14a169873cfe6abc83a39a2df21bc2)

Since it is a complex regex I suggest looking into the unit tests, I tried to cover most of the corner cases

File: adapta/storage/models/aws.py [hunk 2]
```diff
@@ -70,25 +70,30 @@ def from_hdfs_path(cls, hdfs_path: str) -> "S3Path":
         """
         assert hdfs_path.startswith("s3a://"), "HDFS S3 path should start with s3a://"
         uri = urlparse(hdfs_path)
-        parsed_path = uri.path.split("/")
+        parsed_path = uri.path.replace("//", "/").split("/")
```
[the remark this answers is not in the document]
Comment (Andre-Lx-Costa) [reply adapta/storage/models/aws.py]: Included this on the regex check done in __post_init__  done in [cbf15e](https://github.com/SneaksAndData/adapta/pull/453/commits/cbf15e9e7b14a169873cfe6abc83a39a2df21bc2)

File: adapta/storage/models/aws.py [hunk 3]
```diff
@@ -62,6 +58,18 @@ def from_uri(cls, url: str) -> "S3Path":
     path: str
     protocol: str = DataProtocols.S3.value
 
+    def __post_init__(self):
+        if not self.bucket:
+            raise ValueError("Bucket must be defined")
+
+        path_regex = r"^(?![0-9]+\.[0-9]+\.[0-9]+\.[0-9]+$)[a-z0-9]([a-z0-9\-]{1,61}[a-z0-9])?(\/(?!.*(\/\/|\\))([^\/].{0,1022}\/?)?)?$"
```
Comment (george-zubrienko) [inline adapta/storage/models/aws.py:65]: simplify this regex to just check for `//`

File: adapta/storage/models/aws.py [hunk 4]
```diff
@@ -62,6 +58,18 @@ def from_uri(cls, url: str) -> "S3Path":
...
+
+        s3_path_without_prefix = f"{self.bucket}/{self.path}"
```
Comment (george-zubrienko) [inline adapta/storage/models/aws.py:67]: regex should only be applied to `self.path`, bucket is checked on line 62, we do not aim to validate bucket name with this PR, out of scope

File: adapta/storage/models/aws.py [hunk 5]
```diff
@@ -62,6 +58,18 @@ def from_uri(cls, url: str) -> "S3Path":
...
+        match = re.match(path_regex, s3_path_without_prefix)
+
+        if not match:
+            raise ValueError(f"Invalid S3Path provided, must comply with : {path_regex}")
```
Comment (george-zubrienko) [inline adapta/storage/models/aws.py:71]: do not put regex in error message, explain in human language :)

**Review by george-zubrienko — CHANGES_REQUESTED** [review]
Needs work
